### PR TITLE
Better error message on unrecognized command

### PR DESCRIPTION
**Pull request by sbruce**
License: MIT
Signed-off-by: Shaun Bruce <shaun.m.bruce@gmail.com>

### CODE REVIEW

File: commands/cli/parse.go [hunk 1]
```diff
@@ -227,11 +229,61 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
...
+		for _, j := range sortableSuggestions {
+			sFinal = append(sFinal, j.cmd)
+		}
+		return sFinal
+	}
+
 	// if we have more arg values provided than argument definitions,
 	// and the last arg definition is not variadic (or there are no definitions), return an error
 	notVariadic := len(argDefs) == 0 || !argDefs[len(argDefs)-1].Variadic
 	if notVariadic && len(inputs) > len(argDefs) {
```
Comment (whyrusleeping) [inline commands/cli/parse.go]: I wonder if there are any legitimate cases where we would want to print this.. Or if this always means 'you spelled it wrong'.

File: commands/cli/parse.go [hunk 2]
```diff
@@ -227,11 +229,61 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
 		numInputs += 1
 	}
 
+	searchUnknownCmd := func(args []string) []string {
```
Comment (whyrusleeping) [inline commands/cli/parse.go]: maybe move this function out into its own file in the same directory, as well as the sorting code below.

**Comment by sbruce** [comment]
Thanks for the feedback. I moved the code to a new file. I also renamed the **searchUknownCmd** func to **suggestUnknownCmd**. That name sounded a little better.

File: commands/cli/cmd_suggestion.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
+package cli
+
+import (
+	"sort"
+	"strings"
+
+	cmds "github.com/ipfs/go-ipfs/commands"
+	levenshtein "github.com/texttheater/golang-levenshtein/levenshtein"
```
Comment (jbenet) [inline commands/cli/cmd_suggestion.go]: will want to vendor. run 
```
make vendor
```
from root.

---

in the event that godeps barfs at you with all sorts of requests, try `go get` on all the needed packages. 

people report that this works too:
```
godep restore
```
or even setting
```
export GOPATH=`pwd`/Godeps/_workspace:$GOPATH
```

File: commands/cli/parse.go [hunk 4]
```diff
@@ -231,7 +231,15 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
 	// and the last arg definition is not variadic (or there are no definitions), return an error
 	notVariadic := len(argDefs) == 0 || !argDefs[len(argDefs)-1].Variadic
 	if notVariadic && len(inputs) > len(argDefs) {
-		return nil, nil, fmt.Errorf("Expected %v arguments, got %v: %v", len(argDefs), len(inputs), inputs)
+		suggestions := suggestUnknownCmd(inputs, root)
+
+		if len(suggestions) > 1 {
+			return nil, nil, fmt.Errorf("Unknown Command \"%s\"\n\nDid you mean any of these?\n\n\t%s", inputs[0], strings.Join(suggestions, "\n\t"))
+		} else if len(suggestions) > 0 {
+			return nil, nil, fmt.Errorf("Unknown Command \"%s\"\n\nDid you mean this?\n\n\t%s", inputs[0], suggestions[0])
```
Comment (jbenet) [inline commands/cli/parse.go]: maybe just:

```
Did you mean?
<cmd>
```

**Comment by jbenet** [comment]
as someone liable to type with one hand, i find this nice. I'm curious what other people think though and whether this is stuff to put in ipfs, or to leave up to shells?

**Comment by whyrusleeping** [comment]
my golden rule: "what does git do?"

```
whyrusleeping@idril ~> git stats
git: 'stats' is not a git command. See 'git --help'.

Did you mean this?
	status
```

**Comment by whyrusleeping** [comment]
@sbruce if you want, one of us can probably handle the godep vendoring. Its quite the obnoxious process

**Comment by sbruce** [comment]
@whyrusleeping Thanks, but I think I finally figured out godeps.

**Comment by jbenet** [comment]
@sbruce _almost_ there. need to run `make vendor` (which runs: `godep save -r ./...` and rewrites the import paths to point to our workspace)

**Comment by sbruce** [comment]
Okay, make vendor should be fixed.  I also squashed, and rebased with master.

**Comment by whyrusleeping** [comment]
LGTM

**Comment by jbenet** [comment]
@sbruce i should've brought this up. could you write a test for it? if people start depending on this, i want to make sure it works. a sharness test would do. this should get you started:

```
cat <the-below> >test/sharness/t0150-clisuggest.sh
chmod +x test/sharness/t0150-clisuggest.sh
```
```sh
#!/bin/sh

test_description="Test ipfs cli cmd suggest"

. lib/test-lib.sh

test_suggest() {


	test_expect_success "test command fails" '
		test_must_fail ipfs <cmd> >actual
	'

	test_expect_success "test one command is suggested" '
		grep "<cmd1>" actual ||
		fsh cat actual
	'

	test_expect_success "test command fails" '
		test_must_fail ipfs <cmd> >actual
	'

	test_expect_success "test multiple commands are suggested" '
		grep "<cmd1>" actual &&
		grep "<cmd2>" actual ||
		fsh cat actual
	'

}

test_init_ipfs

test_suggest

test_launch_ipfs_daemon

test_suggest

test_kill_ipfs_daemon

test_done
```

**Comment by sbruce** [comment]
A sharness test has been added. Let me know if you think it isn't thorough enough.

**Comment by jbenet** [comment]
@sbruce sorry for the delay. I missed the test update. this LGTM. could you rebase on master? i'd be happy to include it in 0.3.6 if it's rebased today (else 0.3.7).

**Comment by jbenet** [comment]
and thanks! :+1: :smile:

**Comment by sbruce** [comment]
I just rebased to master and pushed the branch. It should hopefully be ready to go.

And, you're welcome. I'm looking forward to helping out more.

**Comment by jbenet** [comment]
thanks @sbruce ! :+1:

**Pull request merged**